src/service/plugins/telephony.js

```diff
@@ -92,20 +92,36 @@ var Plugin = GObject.registerClass({
     }
 
     /**
-     * Restore volume, microphone and media player state (if changed), making
-     * sure to unpause before raising volume.
+     * Restore volume and microphone, plus media player state (if changed and restore enabled).
+     *
+     * If we're going to auto-unpause, make sure to do that first
+     * before raising volume.
+     * Otherwise, still raise the volume (so the user can unpause).
      *
      * TODO: there's a possibility we might revert a media/mixer state set for
      *       another device.
      */
     _restoreMediaState() {
         // Media Playback
-        if (this._mpris)
-            this._mpris.unpauseAll();
+        if (this._mpris) {
+            if (this.settings.get_boolean('after-unpause')) {
+                this._mpris.unpauseAll();
+            } else {
+                /* Even if unpause is disabled, we still need to clear
+                 * the list of which players we're holding paused
+                 *
+                 * TODO: This potentially worsens the multi-device
+                 *       problem, since we may be erasing the player
+                 *       list when another device still needs it.
```

Comment:
As the comment notes, this potentially exacerbates an existing problem that can come up when multiple devices are paired to the local machine. Solving that is beyond the scope of this change.

(The problem is that local player state is global to the extension (or should be), but the actions to manage that state are implemented in the per-device plugins. We may need to take a page from GNOME's power-management, which allows applications to request and clear "inhibit" flags that will prevent various transitions, without providing them any direct control over the powersave state or functionality.

So, instead of the current logic here, something like:
1. A device has an active incoming/outgoing call, for which it is configured  to pause media
2. It informs the MPRIS backend of this fact
3. The backend pauses the media players and stores the device away as the reason for pausing
4. A second device starts a call, and informs the backend
5. The backend adds the second device's request to the list of reasons why playback is paused
6. When either device ends its call, it informs the backend which removes that device from the list
7. When _all_ devices have cleared their requests, regardless of the order in which that happens, the list of reasons why playback can't continue will have become empty.
8. The backend therefore resumes playback. (Assuming the user hasn't disabled that, via the new preference from this PR — which should probably be a global setting rather than a per-device one, if we move to an inhibit-based model).

Managing audio volumes is probably an even trickier juggling act, since the states there aren't as simple as "paused" and "not paused" and multiple devices' requests will have more complex interactions.

Also, open question: If a device disconnects from GSConnect, is its request deleted immediately? Deleted after a grace period (to see if it reconnects)? Deleted only if there are no other requests? Deleted only when there _**are**_ other requests?

Reply:
This sounds pretty reasonable.

As for the last question I don't really know if there's a good answer, but the best answer would probably be to err on the side of leaving the media in its current state (ie. paused). Likewise, if the playback is changed externally that should probably clear any pending "inhibitors" and assume the user has taken the steering wheel.

Reply:
@andyholmes That makes sense. So, if a device drops off unexpectedly, it still gets removed from the inhibit list immediately. (Because we have no idea what its call state is anymore, even if it comes back.) _Buuuut_, playback doesn't resume **even if** it's the last device on the list. At which point, since the list is empty, the MPRIS backend still throws away its stored player state and washes its hands of the whole thing, leaving it to the user to do as they please. (Basically, returning to its default state of not managing any player controls, where it will stay until a new call event comes in and prompts it to start the pause-and-monitor cycle over again.)

Reply:
We could post a notification to the Shell,  too, when a device drops off unexpectedly. (Only in the specific case that it had an active call in progress and it's inhibits were dropped. Otherwise, we're just pestering the user with irrelevancies.)

That way, they'll know why the auto resume "didn't work". (And hopefully not generate any new bug reports.)

Reply:
Hmm, I'd lean towards no notification myself, but I see the argument for it. I'll leave that to you :)

Reply:
@andyholmes Oh no! Am I turning into one of those developers **I'M** always complaining about, the ones who think that _their_ information is important enough to bother the user with?!?

(Like whoever came up with the "Charging complete" notification in Android 4.4... or the idiot that decided it was necessary for a locked and sleeping device to _wake up_ in response to being unplugged from its charger.)

Ouch. Yeah, that's gonna leave an existential crisis.

Reply:
@andyholmes So, to monitor devices with active pause tokens for unexpected disconnects from the MPRIS backend code, is a `Gio.bus_watch_name()` on each device's interface my best option? Or are there convenience functions somewhere I can use to ask, "This ID still connected?" 

...Hrm, actually, that wouldn't work, I'd need to know if it'd disconnected and reconnected at any point as well. `bus_watch_name()` it is!

Reply:
Oh, wait, now I see it. The telephony plugin's `.disconnected()` will be called (well _all_ of the plugins' `.disconnected()`s will be called) if the device drops off, I can have it actively inform the backend that it's unexpectedly going bye-bye.

Reply:
You could also use the pattern from the [component](https://github.com/GSConnect/gnome-shell-extension-gsconnect/blob/master/src/service/components/__init__.js#L10-L16) code, with a `acquire()`/`release()` pair that returns some no-op object with a "refCount" property.

Reply:
@andyholmes I ended up concluding that explicit ref/token _counting_ didn't really buy me anything. I wanted to keep a central list (a `Set`, actually) of all devices with an active request, so its `.size` at any moment represents the count. That way:
1. Any duplicate requests that may come through (for whatever reason) can just be processed normally and require zero special handling, since they won't affect the `.size` of the `Set` the way a new, valid request would.
2. The device plugin instances don't have to hold onto any state at all, for their pause requests — they can statelessly fire off updates to the backend whenever a call event occurs that has some pause-affecting action enabled. The backend tracks the state of device requests right alongside the pause state it's already tracking for the media players.

Reply:
Sounds good, I await your commit so I can have a good look :)